### ISSUE TITLE
unique constraints violated when creating documents version 3.2

### ISSUE THREAD

**Issue by hellomanjack** [arangodb/arangodb#2904]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [x] 3.2
- [ ] self-compiled devel branch

Mode:
- [ ] Cluster
- [x] Single-Server

Storage-Engine:
- [ ] mmfiles
- [x] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

![image](https://user-images.githubusercontent.com/25783698/28723316-f9155ace-73e8-11e7-8333-617d63cecbc3.png)
![image](https://user-images.githubusercontent.com/25783698/28723397-3566aa00-73e9-11e7-9094-69a75e1b7456.png)
![image](https://user-images.githubusercontent.com/25783698/28723467-68c9f370-73e9-11e7-8ad7-b1a1b20f2fc9.png)

I reconfigure the arangod.conf and start the docker. when insert small dates, it works fine.
But when I use the "SORT RAND() Limit 10000" , then memory consumption goes up quickly and the server down.
So I restart the docker. Problems comes,reading documents is ok from the large collection. But create operation is not allowed in that collection. Other collection is fine.

**Comment by Simran-B** [comment]
Sounds similar to what I've encountered when I copied documents from one to another collection:

if the server crashes during execution, some documents will already be copied over to the other collection. If you try to re-run your query, a unique constraint violation will occur because of the already copied documents.

With mmfiles engine, this shouldn't be the case, because of the all-or-nothing fashion of a query. If the server process crashes in the middle of the operation, there will be no documents in the target collection. If the operation succeeds, all documents will be there. During the operation, none will be visible in the target collection (transactional).

With rocksdb engine and document-level locks, it seems like documents copied before a crash remain in the target collection (non-transactional). You can't see new documents while the query executes however (at least using the Web UI).

**Comment by hellomanjack** [comment]
@Simran-B Yes, the server has crashed in the procedure.

**Comment by jsteemann** [comment]
A few comments:

@hellomanjack : The `SORT RAND() LIMIT 10000` will be very quick with the MMFiles engine. It has a special primary index random access method that does not actually sort but pseudo-randomly jump around in the in-memory index.
With the RocksDB engine, `SORT RAND() LIMIT 10000` will be very slow. It will calculate a random value for every document in the collection, then sort all documents by that random value and return the first 10,000. That means it may load the entire collection into memory temporarily. This is somewhat unavoidable, because with RocksDB there is no means to efficiently jump around in the key space nor do we know "good" keys to partition the key space for random jumps.

So `SORT RAND() LIMIT ...` should be avoided with the RocksDB engine if possible, at least for collections with sizes other than very small.

Regarding the actual unique constraint violation: was the collection in question using the autoincrement key generator or the default one?

@Simran-B : If the server crashed somewhere in the middle of copying the data from one collection into the other, then there may be some data already stored in the target collection. If the data is copied via AQL, the AQL transaction may use intermediate commits if the number of writes becomes too big. So some writes may be committed and the overall insert AQL query will not necessarily be atomic. This can be controlled via some transaction and/or query options.

**Comment by hellomanjack** [comment]
@jsteemann I will avoid using the RAND(). I removed the _key,_id then insert the document. And I didn't set the the key pattern . So it should be the default key.

**Comment by jsteemann** [comment]
@hellomanjack : did the collection in question have a special key generator or was it using the default one?
For example,
```js
db._create("test", { keyOptions: { type: "autoincrement" } });
```
generates a collection with an auto-increment key generator, whereas
```js
db._create("test");
```
generates a collection without.
Can you check via `db.<collectionname>.properties()` which key generator was in place for that collection?
Thanks.

**Comment by hellomanjack** [comment]
@jsteemann I create the collection with UI web index.html.

**Comment by jsteemann** [comment]
Ok, that means the default key generator was used. Thanks.

**Comment by dothebart** [comment]
Hi, 
can you retry this with 3.2.3? Thanks.

**Comment by hellomanjack** [comment]
@dothebart I have not found it yet in the 3.2.3.

**Comment by dothebart** [comment]
ok, so do you think we can close this issue then? We can still reopen if you find it again.

**Comment by dothebart** [comment]
Ok, so I'll close this now.

**Comment by maxkernbach** [comment]
Closing this issue as fixed in 3.2.3.
